### ISSUE TITLE
Messages not passing through

### ISSUE THREAD

**Issue by povilasb** [djc/quinn#296]
I've been testing latest master branch (`ab7e228f115dc9759988b846b6a8a3487e34c8bd`) with this test case: https://github.com/povilasb/quinn-tests/blob/fdde6417332096d786c54700b0a622c479a30c88/src/main.rs#L29 which makes one endpoint connect to 20 others and exchanges 1 MB of data and repeats this 15 times.

Sometimes the test passes, but most of the time it gets stuck. Then after doing some debugging it seems like PING frames are flooding the streams. The deepest I got was that [proto::Connection::reset_keep_alive()](https://github.com/djc/quinn/blob/ab7e228f115dc9759988b846b6a8a3487e34c8bd/quinn-proto/src/connection.rs#L776) gets called by [proto::Connection::on_packet_authenticated()](https://github.com/djc/quinn/blob/ab7e228f115dc9759988b846b6a8a3487e34c8bd/quinn-proto/src/connection.rs#L726).

**Comment by Ralith** [comment]
Thanks for the testing! Exploring the dynamics of #284 as a test-case for the issues described here has proved extremely informative. What follows is a detailed summary of what exactly is happening, the likely causes, and what we can do about it.

Tens of concurrent connections attempting to saturate the network link with a nontrivial amount of data generates heavy load. In particular, it easily overflows UDP; on Linux, the default buffer size (for both read and write) appears to be 212992 bytes per socket, which is less than a quarter the size of Quinn's default flow control window for a single stream, and a little over a third the size of 50 times the recommended initial congestion window, which does not account for all overhead. Unlike a TCP server, this test uses Quinn with a single UDP socket, so the limit is shared between all connections, and furthermore each connection must be counted twice: client and server. This produces severe, non-uniformly-distributed packet loss and reordering, thereby exercises a number of otherwise rarely-encountered states. Congestion will cause the congestion window to be reduced, but this is a heuristics process and likely cannot respond quickly enough to the instantaneously overwhelming load of this test.

The bug fixed in #298 was overwhelmingly the most common source of stuck connections, as reordering of the client's first 1-RTT flight with respect to its final handshake flight would consistently cause it to believe it had transmitted all data in the stream while the server was still missing part. With that fixed, I identified a few distinct timeouts:

1. In a release build, both sides of a connection time out.
The handshake completes successfully, but the client becomes blocked on congestion control after transmitting the flight of packets containing the final handshake messages and the first batch of 1-RTT data and receiving few or no acknowledgments. Either no handshake acknowledgments are received at all, or they are received only after the client has deemed the packets lost, forgotten about them, and transmitted replacements that it expects to be acknowledged instead. By the time the period of packet loss has passed, the server has already discarded the cryptographic keys necessary to decode handshake messages, as it has waited a reasonable period of time following the last (re)transmission and has no other way to know whether the client has received its acknowledgments. Because the client is not transmitting data the server can decode, the server times out after transmitting a number of 1-RTT probe packets. When the client stops receiving these probes, it times out in turn.
2. In a release build, the client side times out without the server encountering any error.
The handshake completes successfully, and the client receives acknowledgment of its final handshake flight and proceeds to transmit the full 1MB of data. The server receives and validates the data and closes the connection, but the client does not receive the close message. The client sends a series of probes and receives no response (not even a stateless reset), eventually timing out.
3. In a debug build, both sides of a connection time out partway through transmission of stream data.
The connection is established without issue, and the client rapidly saturates its congestion window. Packets are sent and received in large batches, with large and increasing delays. When the delays exceed the idle timeout, the connection is lost.

Case 3 looks like a straightforward case of insufficient CPU power in debug mode. Increasing the idle timeout is effective at preventing connection loss.

I believe 1 and 2 have the same root cause: transient packet loss causes all packets to be lost during a critical fixed time interval, leading to an unrecoverable state. For case 1 the interval in question is that of the handshake key discard timer, and for case 2 that of the close timer. High CPU load (such as in a debug build) may compound the issue by reducing the frequency with which the UDP receive buffer is drained. While a stateless reset would normally be expected in case 2, our endpoint logic cannot effectively route stateless resets when multiple connections share the same remote address; this may be a bug, but the draft is not clear on the intended behavior here, and in any case stateless resets are not guaranteed.

There's a limited amount we can do to improve behavior in the face of severe packet loss. Under sufficiently bad conditions, timeouts will be the expected behavior of any transport protocol, no matter that the connection is not actually offline. However, we can make some marginal improvements. 

The most direct solution is to simply increase the UDP buffer sizes. On my laptop, I cannot reproduce the malfunction after setting `net.core.rmem_default` and `net.core.rmem_max` to 21299200, representing a TCP-style allocation of a default-sized buffer for each of the 100 effective connections. Increasing the send buffer size may also be needed. Unfortunately, at least on my system, `rmem_max` is set to the same low limit as `rmem_default` (and similar for `wmem`), preventing Quinn from automatically raising it to a more reasonable value. We should document this issue and encourage applications intended to run unprivileged on systems with default settings to distribute connections between multiple endpoints as a work-around. Privileged applications can use `SO_RCVBUFFORCE`/`SO_SNDBUFFORCE` to disregard the limit.

Case 1 and 2 could be further mitigated by significantly increasing their respective timeouts. The QUIC draft does not currently recommend a specific strategy for the handshake key discard timeout, so it is plausible that our current choice is suboptimal. Quinn currently uses the minimum permitted timeouts, but implementations are permitted to increase both as they see fit. This has drawbacks: a longer key discard timer very slightly weakens security, and a longer close timer will delay recovery of resources (though we could take measures to reduce resources consumed by connections in this state) and may delay application shutdown. These are minor, so a moderate increase in default timeouts is a possibility. A more general solution would be to expose tuning factors. That said, relying on increased timeout periods to work around packet loss due to buffer overflow is not ideal; lost packets are wasted bandwidth, and even if congestion control adapts, throughput will be artificially limited.

**Comment by Ralith** [comment]
Interestingly, Windows is more relaxed in this regard, so we may be able to scale up the receive buffer size as needed there, perhaps even dynamically.

**Comment by povilasb** [comment]
Great info, thanks!

I've tried increasing socket limits, but that didn't help :/
```
sysctl net.core.rmem_max=21299200
net.core.rmem_max = 21299200
```
I'm still seeing a flood of PING packets.
Note that in this test case https://github.com/povilasb/quinn-tests/blob/fdde6417332096d786c54700b0a622c479a30c88/src/main.rs#L27 I'm using a single UDP socket per `Node` and then make each `Node` connect to another `Node`. Eventually each UDP socket executes 19 connections and transmits just 19 MB of data. But even in this case some of the connections are stuck (~17 out of 19 connections successfully transfer the data). Given that PING packets are continuously scheduled, I believe there's some bug disregard to socket RX buffers being full.

**Comment by povilasb** [comment]
Turns out there's no issues with PING packets. There was a problem with my tests: I had keep alive interval set to 1 second and kept all completed connections in a Vec which resulted in sending 400 PING frames every second..

Anyway, but still not all packets get delivered even when socket TX/RX buffer sizes are increased.

**Comment by Ralith** [comment]
> I've tried increasing socket limits, but that didn't help :/

You must set `rmem_default` as well as `rmem_max` for this to have any effect, and should probably set `wmem_*` as well.

> Anyway, but still not all packets get delivered

Do you mean "not all streams get finished?"

> Personally I thought that the congestion control should kick in and allow each connection/stream to keep moving slowly but forward till finished.

Congestion control must observe loss to respond to it, so it will always take time to react. When loss goes abruptly from zero to massive, it may not be possible to recover within the duration of one of those critical time intervals. Transient loss outside of those periods should always be recovered from.

I've pushed a modified version of your test to [the `debugging` branch](https://github.com/djc/quinn/tree/debugging), demonstrating how to capture logs from individual connections. If increasing the default buffer sizes doesn't resolve the issue, could you do something similar here and provide the trace-level logs for the connections that fail to complete? If they don't actually time out, you could e.g. print the logs of connections that haven't completed after a certain amount of time has passed.

**Comment by Ralith** [comment]
FWIW, quinn-tests master completes successfully every time I run it after increasing buffer sizes, even in debug builds, even when I increase the number of connections to 40 (though you have to keep an eye on memory use, since of course n^2 connections requires n^2 memory.

**Comment by nbaksalyar** [comment]
> quinn-tests master completes successfully every time I run it after increasing buffer sizes

That did the trick for me too. But still, with the default buffer sizes (~200KB) it seems like it never finishes. I've tried running the same example over a prolonged period of time (like 10 mins.) and it still seems like being stuck. I guess even with small buffers it should resolve at least in _some_ time?

**Comment by Ralith** [comment]
> I guess even with small buffers it should resolve at least in some time?

As outlined in [my writeup](https://github.com/djc/quinn/issues/296#issuecomment-480651069), there are two critical periods where complete packet loss covering the period in question (as severely undersized buffer can cause) will cause a connection to enter an unrecoverable state. Normally this will lead to an idle timeout, but the quinn-tests code disables those. We can make those periods larger, and thus less likely to be fully eclipsed by a period of congestion, but we can't make them infinite, and using correctly sized buffers will substantially improve performance regardless.

These issues arise when we're receiving data faster than we can read it (or vis versa), so reduced bandwidth requirements (e.g. due to running over the WAN; dumping bulk data over loopback or a datacenter network are worst cases here), increased CPU power, and optimization in Quinn should all improve matters.

**Comment by Ralith** [comment]
I think we can resolve case 1 by dramatically increasing the handshake key discard timeout without any significant drawbacks, and case 2 can be handled in application code by treating time-outs similarly to the connection being closed by the peer.

**Comment by povilasb** [comment]
> Do you mean "not all streams get finished?"

Yes that's what I meant :)

Indeed, setting the default rmem limit makes the streams finish. And setting the timeout does terminate the connection :+1:

**Comment by povilasb** [comment]
> Congestion control must observe loss to respond to it, so it will always take time to react. When loss goes abruptly from zero to massive, it may not be possible to recover within the duration of one of those critical time intervals. 

Hmm, I've been reading about packet loss detection https://tools.ietf.org/html/draft-ietf-quic-recovery-19#section-6.1.2

```
The time threshold is computed as kTimeThreshold *
   max(SRTT, latest_RTT).  If packets sent prior to the largest
   acknowledged packet cannot yet be declared lost, then a timer SHOULD
   be set for the remaining time.

kTimeThreshold = 9/8
```

So based on this info seems like a packet should be considered lost roughly after 1 RTT. During the tests I'm running on localhost 1 RTT should be really low, which in turn should detect packet loss really quickly and trigger congestion control.

```
When a loss is detected,
   NewReno halves the congestion window and sets the slow start
   threshold to the new congestion window.
```
https://tools.ietf.org/html/draft-ietf-quic-recovery-19#section-7.3

Am I missing something here? :thinking:

**Comment by Ralith** [comment]
> So based on this info seems like a packet should be considered lost roughly after 1 RTT

This is inaccurate. Packets are deemed lost by the time threshold logic if, were they to be received, they would be *reordered across* a period larger than the time threshold. If you're not receiving anything at all, then you're not detecting reordering, so it does not apply. Deeming packets lost after a single RTT would lead to poor performance, as routine variations in latency would cause spurious loss detection and receivers would not be able to batch acknowledgements. In fact, the standard QUIC logic never deems a packet lost until a later packet has been acknowledged; if no ACKs are being received, then the sender focuses on eliciting them with probe packets.

There is certainly room to be cleverer about congestion control; this is an active research area, and I'm keeping an eye on projects like google's [BBR](https://github.com/google/bbr). Ultimately I'd like to make congestion control pluggable so people can easily experiment in external code. That said, the standard logic should generally be Good Enough; if it turns out to have serious shortcomings we can raise that with the IETF.

There's one remaining bit of strangeness here where case 2 times out when it should be seeing stateless reset packets (at least as of #301), but probably that deserves to be pulled out into its own issue if I can't track it down in the next day or two. Great to hear things seem to be working for you!
